File: python/identitykit/identity.py
"""The AgentIdentity document: a signed, portable description of an agent and who
operates it. JSON-friendly dict, signed with a detached Ed25519 proof.

Key control differs by identifier method, and that difference is the whole point:

  did:key  the id *is* a public key. The proof must be made by that exact key.
           Self-certifying and offline-verifiable; trust nothing but the key.
  did:web  the id is a domain. The document is fetched over TLS from that domain,
           and the proof must be made by a key the document itself lists. Trust
           is anchored in DNS/TLS, not the key alone.
"""

from __future__ import annotations

from typing import Dict, List, Optional

from . import signing
from .errors import InvalidIdentity

SCHEMA = "https://identitykit.dev/schema/agent-identity/v0"

VALID_PURPOSES = {"controller", "auth", "mandate-issuer", "witness-signer", "assertion"}
VALID_BINDING_STANDARDS = {"did", "fido", "ap2", "eudi", "vc"}
VALID_OPERATOR_TYPES = {"org", "person"}


def _require(cond: bool, message: str, field: Optional[str] = None) -> None:
    if not cond:
        raise InvalidIdentity(message, field=field)


def build_identity(
    id: str,
    operator: Dict,
    keys: List[Dict],
    *,
    capabilities: Optional[List[str]] = None,
    bindings: Optional[List[Dict]] = None,
    created: Optional[str] = None,
) -> Dict:
    """Assemble an unsigned identity document. Validates structure; sign with
    sign_identity()."""
    doc = {
        "@schema": SCHEMA,
        "id": id,
        "operator": operator,
        "keys": keys,
        "capabilities": list(capabilities or []),
        "bindings": list(bindings or []),
    }
    if created is not None:
        doc["created"] = created
    validate(doc)
    return doc


def validate(doc: Dict) -> Dict:
    """Structural validation. Raises InvalidIdentity; returns the document."""
    _require(isinstance(doc, dict), "identity must be an object")
    _require(isinstance(doc.get("id"), str) and doc["id"], "id is required", "id")
    method = doc["id"].split(":")
    _require(len(method) >= 3 and method[0] == "did" and method[1] in ("key", "web"),
             "id must be a did:key or did:web identifier", "id")

    op = doc.get("operator")
    _require(isinstance(op, dict), "operator is required", "operator")
    _require(isinstance(op.get("name"), str) and op["name"], "operator.name is required", "operator.name")
    _require(op.get("type") in VALID_OPERATOR_TYPES,
             f"operator.type must be one of {sorted(VALID_OPERATOR_TYPES)}", "operator.type")

    keys = doc.get("keys")
    _require(isinstance(keys, list) and len(keys) >= 1, "at least one key is required", "keys")
    seen_ids = set()
    for i, k in enumerate(keys):
        _require(isinstance(k, dict), f"keys[{i}] must be an object", "keys")
        _require(isinstance(k.get("id"), str) and k["id"], f"keys[{i}].id is required", "keys")
        _require(k["id"] not in seen_ids, f"duplicate key id {k['id']!r}", "keys")
        seen_ids.add(k["id"])
        _require(isinstance(k.get("public_key"), str) and k["public_key"],
                 f"keys[{i}].public_key is required", "keys")
        purposes = k.get("purpose")
        plist = purposes if isinstance(purposes, list) else [purposes]
        _require(all(p in VALID_PURPOSES for p in plist),
                 f"keys[{i}].purpose must be from {sorted(VALID_PURPOSES)}", "keys")
        try:
            signing.unb64(k["public_key"])
        except Exception:
            raise InvalidIdentity(f"keys[{i}].public_key is not valid base64", field="keys")

    bindings = doc.get("bindings", [])
    _require(isinstance(bindings, list), "bindings must be a list", "bindings")
    for i, b in enumerate(bindings):
        _require(isinstance(b, dict) and b.get("standard") in VALID_BINDING_STANDARDS,
                 f"bindings[{i}].standard must be from {sorted(VALID_BINDING_STANDARDS)}", "bindings")
        _require(isinstance(b.get("ref"), str) and b["ref"], f"bindings[{i}].ref is required", "bindings")

    _require(isinstance(doc.get("capabilities", []), list), "capabilities must be a list", "capabilities")
    return doc


def sign_identity(doc: Dict, seed: bytes, *, key_id: Optional[str] = None) -> Dict:
    """Attach a detached Ed25519 proof. The document (without proof) is what gets
    signed, so re-verification is exact."""
    validate(doc)
    body = {k: v for k, v in doc.items() if k != "proof"}
    signed = dict(body)
    signed["proof"] = signing.make_proof(body, seed, key_id=key_id)
    return signed


def verify_identity(doc: Dict) -> bool:
    """Structural validity + proof validity + key control for the id method.
    Never raises; returns False on any failure."""
    try:
        validate(doc)
        proof = doc.get("proof")
        if not isinstance(proof, dict):
            return False
        body = {k: v for k, v in doc.items() if k != "proof"}
        if not signing.verify_proof(body, proof):
            return False

        proof_key = signing.unb64(proof["public_key"])
        method = doc["id"].split(":")[1]

        if method == "key":
            # Self-certifying: the proof key must BE the key in the identifier.
            return signing.public_from_did_key(doc["id"]) == proof_key

        if method == "web":
            # Domain-anchored: the proof key must be one the document lists with
            # a controlling purpose. (Domain/TLS trust is the resolver's job.)
            for k in doc["keys"]:
                purposes = k.get("purpose")
                plist = purposes if isinstance(purposes, list) else [purposes]
                if not ({"controller", "auth"} & set(plist)):
                    continue
                if signing.unb64(k["public_key"]) == proof_key:
                    return True
            return False

        return False
    except Exception:
        return False

File: typescript/src/identity.ts
/** The AgentIdentity document: signed, portable, with method-specific key control.
 *  did:key is self-certifying; did:web is domain-anchored. */

import * as signing from "./signing.ts";
import { InvalidIdentity } from "./errors.ts";

export const SCHEMA = "https://identitykit.dev/schema/agent-identity/v0";

export const VALID_PURPOSES = new Set(["controller", "auth", "mandate-issuer", "witness-signer", "assertion"]);
export const VALID_BINDING_STANDARDS = new Set(["did", "fido", "ap2", "eudi", "vc"]);
export const VALID_OPERATOR_TYPES = new Set(["org", "person"]);

export interface Key {
  id: string;
  purpose: string | string[];
  public_key: string;
}
export interface Operator {
  name: string;
  type: "org" | "person";
  [k: string]: unknown;
}
export interface Binding {
  standard: string;
  ref: string;
}
export interface AgentIdentity {
  "@schema": string;
  id: string;
  operator: Operator;
  keys: Key[];
  capabilities: string[];
  bindings: Binding[];
  created?: string;
  proof?: signing.Proof;
}

function req(cond: boolean, message: string, field?: string): void {
  if (!cond) throw new InvalidIdentity(message, field);
}

const purposesOf = (k: Key): string[] => (Array.isArray(k.purpose) ? k.purpose : [k.purpose]);

export interface BuildOpts {
  capabilities?: string[];
  bindings?: Binding[];
  created?: string;
}

export function buildIdentity(id: string, operator: Operator, keys: Key[], opts: BuildOpts = {}): AgentIdentity {
  const doc: AgentIdentity = {
    "@schema": SCHEMA,
    id,
    operator,
    keys,
    capabilities: [...(opts.capabilities ?? [])],
    bindings: [...(opts.bindings ?? [])],
  };
  if (opts.created !== undefined) doc.created = opts.created;
  validate(doc);
  return doc;
}

export function validate(doc: AgentIdentity): AgentIdentity {
  req(!!doc && typeof doc === "object", "identity must be an object");
  req(typeof doc.id === "string" && !!doc.id, "id is required", "id");
  const m = doc.id.split(":");
  req(m.length >= 3 && m[0] === "did" && (m[1] === "key" || m[1] === "web"), "id must be did:key or did:web", "id");

  const op = doc.operator;
  req(!!op && typeof op === "object", "operator is required", "operator");
  req(typeof op.name === "string" && !!op.name, "operator.name is required", "operator.name");
  req(VALID_OPERATOR_TYPES.has(op.type), "operator.type must be org|person", "operator.type");

  req(Array.isArray(doc.keys) && doc.keys.length >= 1, "at least one key is required", "keys");
  const seen = new Set<string>();
  doc.keys.forEach((k, i) => {
    req(!!k && typeof k === "object", `keys[${i}] must be an object`, "keys");
    req(typeof k.id === "string" && !!k.id, `keys[${i}].id is required`, "keys");
    req(!seen.has(k.id), `duplicate key id ${k.id}`, "keys");
    seen.add(k.id);
    req(typeof k.public_key === "string" && !!k.public_key, `keys[${i}].public_key is required`, "keys");
    req(purposesOf(k).every((p) => VALID_PURPOSES.has(p)), `keys[${i}].purpose invalid`, "keys");
    try {
      const decoded = signing.unb64(k.public_key);
      req(decoded.length > 0, `keys[${i}].public_key empty`, "keys");
    } catch {
      throw new InvalidIdentity(`keys[${i}].public_key is not valid base64`, "keys");
    }
  });

  const bindings = doc.bindings ?? [];
  req(Array.isArray(bindings), "bindings must be a list", "bindings");
  bindings.forEach((b, i) => {
    req(!!b && VALID_BINDING_STANDARDS.has(b.standard), `bindings[${i}].standard invalid`, "bindings");
    req(typeof b.ref === "string" && !!b.ref, `bindings[${i}].ref is required`, "bindings");
  });
  req(Array.isArray(doc.capabilities ?? []), "capabilities must be a list", "capabilities");
  return doc;
}

function body(doc: AgentIdentity): Record<string, unknown> {
  const { proof, ...rest } = doc;
  void proof;
  return rest;
}

export function signIdentity(doc: AgentIdentity, seed: Buffer, keyId?: string): AgentIdentity {
  validate(doc);
  const { proof: _omit, ...b } = doc;
  void _omit;
  return { ...b, proof: signing.makeProof(b, seed, keyId) };
}

export function verifyIdentity(doc: AgentIdentity): boolean {
  try {
    validate(doc);
    const proof = doc.proof;
    if (!proof || typeof proof !== "object") return false;
    const b = body(doc);
    if (!signing.verifyProof(b, proof)) return false;

    const proofKey = signing.unb64(proof.public_key);
    const method = doc.id.split(":")[1];

    if (method === "key") {
      return signing.publicFromDidKey(doc.id).equals(proofKey);
    }
    if (method === "web") {
      for (const k of doc.keys) {
        const ps = purposesOf(k);
        if (!ps.includes("controller") && !ps.includes("auth")) continue;
        if (signing.unb64(k.public_key).equals(proofKey)) return true;
      }
      return false;
    }
    return false;
  } catch {
    return false;
  }
}
